### ISSUE TITLE
Signature for this request is not valid

### ISSUE THREAD

**Issue by yarnik1** [taskula/binance-perl-api#8]
simple request to view the balance
`use Binance::API;
my $api = Binance::API->new(
    apiKey    => 'myapiKey',
    secretKey => 'mysecretKey',
);
$test = $api->account( );
$mess.=pre(Dumper \$test);`

I see an error
`'_content' => '{"code":-1022,"msg":"Signature for this request is not valid."}',`

no problem with the ticker request `my $ticker = $api->ticker( symbol => 'ETHBTC' );`

and simple code work too:
`use Digest::SHA qw(hmac_sha256_hex);
use Time::HiRes qw(time);
$api_key = 'myapiKey';
$api_secret = 'mysecretKey';
$url = URI->new('https://api.binance.com/api/v3/account');
$data = "timestamp=" . int(time() * 1000);
$signature = uc(hmac_sha256_hex($data, $api_secret));
$ua = LWP::UserAgent->new();
$ua->default_header('X-MBX-APIKEY' => $api_key);
$response = $ua->get( "$url?$data&signature=$signature" );`

**Comment by taskula** [comment]
I cannot replicate this issue, `account()` works for me. More info needed. You can use `BINANCE_API_DEBUG=1` environment variable with your script to print out the request.

I would look at https://github.com/taskula/binance-perl-api/blob/7f79bec83276cbb579dddda1cb289be6b3f6036d/lib/Binance/API/Request.pm#L171-L174 and see if anything strange happens here.

**Comment by yarnik1** [comment]
How can I do it?
I have tried it, but no result:
`use Binance::API;
$ENV{BINANCE_API_DEBUG} = 1;
my $api = Binance::API->new(
    apiKey    => 'myapiKey',
    secretKey => 'mysecretKey',
);
my $test = $api->account(  );
$mess.=pre(Dumper \$logger, \$test );`

**Comment by taskula** [comment]
You can pass the environment variable directly in your shell, just run your script by executing the following command `BINANCE_API_DEBUG=1 perl yourscript.pl`.

But right now it is not printing much useful information other than the full request path.

Judging by the error code you are getting, you need to dig into the file I mentioned and see how the signature is generated.

Is this your full code? Could you format your code properly by using triple ticks https://docs.github.com/en/github/writing-on-github/creating-and-highlighting-code-blocks#fenced-code-blocksand so that it is easier to read. Are you using 
```
use strict;
use warnings;
```
?

**Comment by yarnik1** [comment]
```#!/usr/bin/perl
use strict;
use warnings;
use Data::Dumper;
use Binance::API;
my $api = Binance::API->new(
	apiKey    => '...',
	secretKey => '...',
);
my $test = $api->account();
print (Dumper \$test);
```
\# BINANCE_API_DEBUG=1 perl binance_test.cgi
```
[Binance::API::Request::_exec] New request: https://api.binance.com/api/v3/account?signature=7125e5a295a90569.....181ed9c07b&timestamp=1600181348260&recvWindow=5000 at /usr/local/share/perl5/Binance/API/Request.pm line 87.
[Binance::API::Request::_exec] Bad Request at /usr/local/share/perl5/Binance/API/Request.pm line 104.
$VAR1 = \bless( {
                   '_protocol' => 'HTTP/1.1',
                   '_content' => '{"code":-1022,"msg":"Signature for this request is not valid."}',
                   ....
```
and works wonderfully with $api->cancel_order();
```
use strict;
use warnings;
use Data::Dumper;
use Binance::API;
my $api = Binance::API->new(
	apiKey    => '...',
	secretKey => '...',
);
my $test = $api->cancel_order(symbol => 'BTTUSDT', orderId=>'115833840');
print (Dumper \$test);
```
\# BINANCE_API_DEBUG=1 perl binance_test.cgi
```
[Binance::API::Request::_exec] New request: https://api.binance.com/api/v3/order at /usr/local/share/perl5/Binance/API/Request.pm line 87.
$VAR1 = \{
            'symbol' => 'BTTUSDT',
            'origQty' => '1.00000000',
            'status' => 'CANCELED',
            'orderListId' => -1,
            'origClientOrderId' => '...',
            'timeInForce' => 'GTC',
            'clientOrderId' => '...',
            'side' => 'SELL',
            'orderId' => 115833840,
            'executedQty' => '0.00000000',
            'cummulativeQuoteQty' => '0.00000000',
            'type' => 'LIMIT',
            'price' => '0.00100000'
          };
```

**Comment by yarnik1** [comment]
Eureka!
```
# timestamp should be the first
$url.= "?timestamp=$data{'timestamp'}&signature=$data{'signature'}";
# error if at the end
$url.= "?signature=$data{'signature'}&timestamp=$data{'timestamp'}";

**Comment by taskula** [comment]
Thank you for the update, nice to see progress.

But I find this strange, because Binance REST API documentation states that
``Parameters may be sent in any order.`` at https://github.com/binance-exchange/binance-official-api-docs/blob/ede89576a345e4d84e4397045136d9371b9ca168/rest-api.md

Additionally, I was debugging Binance::API::Request and found out that my signature parameter comes after the timestamp (as in your second case) and still works.

Can you create a failing test in https://github.com/taskula/binance-perl-api/blob/master/t/Binance/API/Request.t to prove this?

If you don't know how, I can help you with it.

**Comment by yarnik1** [comment]
I don't know how, can you help me with it.

**Comment by taskula** [comment]
Yes. It is not trivial because in order to run tests against live Binance API we need an API key to perform requests of an actual account.

I don't know if Binance provides such test accounts but we can look into it. If not, we could create a dummy account with empty wallets just for this project. We then need to mock the hashing algorithm in a way that allows us to set the order of parameters any way we desire. This can be achieved with ``Test::MockModule`` test library and overriding the ``Binance::API::Request::_init()`` method in a way that lets us control the sorting of parameters. Even better we should refactor the ``_init()`` method so that we can mock smaller parts of it, such as only the query parameter generation section.

Anyways, I created a small script to demonstrate the issue and it appears there indeed is something with the order of parameters:

```
use strict;
use warnings;
use Data::Dumper;
use Digest::SHA qw( hmac_sha256_hex );
use JSON;
use Time::HiRes;
use Binance::API;

my $api = Binance::API->new(
        apiKey    => '...',
        secretKey => '...',
);
my $base_url = "https://api.binance.com/api/v3/account?";
my $timestamp = int Time::HiRes::time * 1000;
# order that works
my $url .= $base_url."recvWindow=5000";
$url .= "&timestamp=$timestamp";
$url .= "&signature=".hmac_sha256_hex("recvWindow=5000&timestamp=$timestamp", $api->{ua}->{secretKey});

print "$url\n";
print $api->{ua}->_exec("get", $url, X_MBX_APIKEY => $api->{ua}->{apiKey}) . "\n";

$url = "&recvWindow=5000";
$url .= "&timestamp=$timestamp";
$url = $base_url."signature=".hmac_sha256_hex("recvWindow=5000&timestamp=$timestamp", $api->{ua}->{secretKey}) . $url;

print "$url\n";
print $api->{ua}->_exec("get", $url, X_MBX_APIKEY => $api->{ua}->{apiKey}) . "\n";
```

The first execution works for me but the second doesn't. So it appears the signature should be at the very end.

I will continue to take a closer look at this tomorrow, but thanks for reporting the findings.

Meanwhile you are welcome to send any pull requests.

**Comment by yarnik1** [comment]
because the signature always comes at the end and must sign all the parameters that come before it

**Comment by taskula** [comment]
Hi,

Sorry I wasn't able to help you a month ago.

Anyway, can you test this branch https://github.com/taskula/binance-perl-api/tree/issue_8 and see if it fixes your issues?

It forces signature into the end of parameter list.

**Comment by taskula** [comment]
I tested it myself and pushed it to master. Could you check if the master branch has solved your issue?

**Comment by yarnik1** [comment]
Sorry, but I have server with Centos with installed old version binance-perl-api.
When the new version will be available for the update, I do not know.
`yum updete` gives nothing yet.
I made this solution for myself:
```
sub new {
	my ($method, $url, $params) = @_;
	my $data = $params->{'data'};
	my $ua = LWP::UserAgent->new();
	my $response;

	$data->{'timestamp'} = int(time() * 1000) if $params->{'signed'};
	my $param = keys %$data ? join "&", map { $_ . '=' . $data->{$_} } sort {$b cmp $a} keys %$data : '';

	if ($params->{'signed'}) {
		$ua->default_header('X-MBX-APIKEY' => $apikey);
		$param.= '&signature=' . uc(hmac_sha256_hex($param, $apisecret));
	}

	$url.= '?' . $param if $param;
	$response = $ua->$method( $url );
        # .....
}
```
And... probably sorting is not needed here (this was one of the experimental options)

**Comment by taskula** [comment]
> When the new version will be available for the update, I do not know.

You should be able to update to the latest version by running the command `cpan Binance::API`.

> yum updete gives nothing yet.

I have not packaged the module, and am not aware that anyone else packaged it either, so it shouldn't be possible to update it via your package manager anyway. How did you install it in the first place?

**Comment by taskula** [comment]
Closing due to inactivity. Fix introduced in ee11308111a7feb3fb032e2d0d78f9b259740af3.
